extensions/eclipse-che-theia-plugin-remote/src/node/che-plugin-scanner.ts

```diff
@@ -0,0 +1,53 @@
+/**********************************************************************
+ * Copyright (c) 2020 Red Hat, Inc.
+ *
+ * This program and the accompanying materials are made
+ * available under the terms of the Eclipse Public License 2.0
+ * which is available at https://www.eclipse.org/legal/epl-2.0/
+ *
+ * SPDX-License-Identifier: EPL-2.0
+ ***********************************************************************/
+
+import * as path from 'path';
+
+import { PluginPackage, SnippetContribution, getPluginId } from '@theia/plugin-ext/lib/common';
+
+import { PluginUri } from '../common/plugin-uri';
+import { TheiaPluginScanner } from '@theia/plugin-ext/lib/hosted/node/scanners/scanner-theia';
+import { VsCodePluginScanner } from '@theia/plugin-ext-vscode/lib/node/scanner-vscode';
+
+export class CheTheiaPluginScanner extends TheiaPluginScanner {
+  protected readSnippets(pck: PluginPackage): SnippetContribution[] | undefined {
+    return readSnippets(pck);
+  }
+}
+
+export class CheVsCodePluginScanner extends VsCodePluginScanner {
+  protected readSnippets(pck: PluginPackage): SnippetContribution[] | undefined {
+    return readSnippets(pck);
+  }
+}
+
+/**
+ * Override reading snippet contributions to use plugin resource uris instead of
+ * file uris.
+ */
+function readSnippets(pck: PluginPackage): SnippetContribution[] | undefined {
+  if (!pck.contributes || !pck.contributes.snippets) {
+    return undefined;
+  }
+  const result: SnippetContribution[] = [];
+  for (const contribution of pck.contributes.snippets) {
+    if (contribution.path) {
+      const absolutePath = path.join(pck.packagePath, contribution.path);
+      const normalizedPath = path.normalize(absolutePath);
+      const relativePath = path.relative(pck.packagePath, normalizedPath);
+      result.push({
+        language: contribution.language,
+        source: pck.displayName || pck.name,
+        uri: PluginUri.createUri(getPluginId(pck), relativePath),
```

Comment:
we should just simply make a SnippetContributionProviderURI or SnippetContributionProvider singleton upstream and then just overrides how we create URI downstream
Here there are too many overrides

Reply:
What do you mean, there are too many overrides? It's a single method. If we extract the various contribution providers upstream, we would logically have to to it for all kinds of contributions. That may or may not be desirable, but one of the reasons we're doing this downstream is that the work was deemed urgent.

Reply:
I mean here that there are too many classes being extended (and we may miss some of them)
while we're only interested on delegating the contribution uri.

> That may or may not be desirable, but one of the reasons we're doing this downstream is that the work was deemed urgent.

https://github.com/eclipse/che/issues/16502 is opened since march
I think that adding a new simple method in upstream theia to handle the contribution can be accepted in a couple of days as it's a very minor change.

Reply:
As far as I understand it, this issue is not reproducible in upstream theia. In the interest of saving time, I suggest we keep the fix downstream: not only because the benefit for theia is low, but also because the holidays are coming and getting the PR merged on time is not guaranteed. We can always iterate later if the need for such methods in theia arises.

Reply:
'sdon't think it would be a problem to have something merge upstream if it's only an helper to get an URI for a plugin resource. Upstream theia having implementation with File and downstream with something else.
Discussing it is taking more time than having proposed a simple tiny PR url upstream.

Also I see two different ways being implemented for getting resources for a plug-in this one and https://github.com/eclipse/che-theia/pull/946 which is kind of colliding.

Reply:
I asked explicitly on a Theia dev call whether separation of plugin host and back end on different machines was something that should be supported in Theia and the answer was no. Hence the downstream effort instead of pursuing https://github.com/eclipse-theia/theia/pull/8468

Reply:
And again, we can push for this refactoring upstream any time later, but we already have all the necessary parts to fix the issue.

Reply:
It's different. It's not to have a plug-in-resource: protocol upstream, just to have something than convert a plug-in path to an URI by using File and that can be replaced in Che (with protocols, etc)

Reply:
I agree with @benoitf. Here using inheritance and overrides could be quite hard to maintain in the future. It is overriding a complete method and copying some parts. So somehow, upstream someone change that line https://github.com/eclipse-theia/theia/blob/bd25d737742a0974b911da5d3a5c978d950fc64e/packages/monaco/src/browser/monaco-snippet-suggest-provider.ts#L146, it would not be reflected downstream and could cause bugs that are hard to track.